lib/swoosh/adapters/dyn.ex

```diff
@@ -0,0 +1,102 @@
+  defmodule Swoosh.Adapters.Dyn do
+  @moduledoc ~S"""
+  An adapter that sends email using the Dyn API.
+
+  For reference: [Dyn API docs](https://help.dyn.com/email-rest-methods-api/sending-api/)
+
+  ## Example
+
+      # config/config.exs
+      config :sample, Sample.Mailer,
+        adapter: Swoosh.Adapters.Dyn,
+        api_key: "my-api-key",
+        domain: "avengers.com"
+
+      # lib/sample/mailer.ex
+      defmodule Sample.Mailer do
+        use Swoosh.Mailer, otp_app: :sample
+      end
+  """
+  use Swoosh.Adapter, required_config: [:api_key]
+
+  alias Swoosh.Email
+  import Swoosh.Email.Render
+
+  @base_url     "https://emailapi.dynect.net"
+  @api_endpoint "rest/json/send"
+
+  def deliver(%Email{} = email, config \\ []) do
+    headers = prepare_headers(email, config)
+    url = [base_url(config), "/", @api_endpoint]
+
+    body = %{}
+      |> prepare_from(email)
+      |> prepare_to(email)
+      |> prepare_subject(email)
+      |> prepare_html(email)
+      |> prepare_text(email)
+      |> prepare_cc(email)
+      |> prepare_bcc(email)
+      |> prepare_reply_to(email)
+      |> add_auth_token(config[:api_key])
+      |> encode_body
+
+    case :hackney.post(url, headers, body, [:with_body]) do
+      {:ok, 200, _headers, body} ->
+        {:ok, Poison.decode!(body)["response"]["message"]}
+      {:ok, 404, _headers, body} ->
+        {:error, "Not found"}
+      {:ok, 503, _headers, body} ->
+        {:error, "Service Unavailable"}
+      {:ok, code, _headers, body} ->
+        {:error, "Error: #{inspect(body)}"}
+      {:error, reason} ->
+        {:error, reason}
+    end
+  end
+
+  defp base_url(config), do: config[:base_url] || @base_url
+
+  defp prepare_headers(email, config) do
+    [{"User-Agent", "swoosh/#{Swoosh.version}"},
+     {"Content-Type", content_type(email)}]
+  end
+
+  defp content_type(%{attachments: []}), do: "application/x-www-form-urlencoded"
+  defp content_type(%{}), do: "multipart/form-data"
+
+  defp prepare_from(body, %{from: from}), do: Map.put(body, :from, render_recipient(from))
+
+  defp prepare_to(body, %{to: to}), do: Map.put(body, :to, render_recipient(to))
+
+  defp prepare_reply_to(body, %{reply_to: nil}), do: body
+  defp prepare_reply_to(body, %{reply_to: {_name, address}}), do: Map.put(body, :replyto, address)
+
+  defp prepare_cc(body, %{cc: []}), do: body
+  defp prepare_cc(body, %{cc: ccs}) do
+    ccs
+    |> Enum.with_index(1)
+    |> Enum.map(fn {cc, counter} -> {String.to_atom("cc[#{counter}]"), render_recipient(cc)} end)
```

Comment:
this and bcc don't need to be atoms, and shouldn't. if someone cc 10k people, this would create 10k atoms that will never be garbage collected.